src/Tempest/Http/src/Routing/Construction/RouteTreeNode.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Http\Routing\Construction;
+
+use Tempest\Http\Route;
+
+/**
+ * @internal
+ */
+final class RouteTreeNode
+{
+    /** @var array<string, RouteTreeNode> */
+    private array $staticPaths = [];
+
+    /** @var array<string, RouteTreeNode> */
+    private array $dynamicPaths = [];
+
+    private ?MarkedRoute $leaf = null;
+
+    private function __construct(
+        public readonly RouteTreeNodeType $type,
+        public readonly ?string $segment = null
+    ) {
+    }
+
+    public static function createRootRoute(): self
+    {
+        return new self(RouteTreeNodeType::Root);
+    }
+
+    public static function createParameterRoute(string $regex): self
+    {
+        return new self(RouteTreeNodeType::Parameter, $regex);
+    }
+
+    public static function createStaticRoute(string $name): self
+    {
+        return new self(RouteTreeNodeType::Static, $name);
+    }
+
+    public function addPath(array $pathSegments, MarkedRoute $markedRoute): void
+    {
+        if ($pathSegments === []) {
+            if ($this->leaf !== null) {
+                throw new DuplicateRouteException($markedRoute->route);
+            }
+
+            $this->leaf = $markedRoute;
+
+            return;
+        }
+
+        $currentPathSegment = array_shift($pathSegments);
+
+        $regexPathSegment = self::convertDynamicSegmentToRegex($currentPathSegment);
+
+        if ($currentPathSegment !== $regexPathSegment) {
+            $node = $this->dynamicPaths[$regexPathSegment] ??= self::createParameterRoute($regexPathSegment);
+        } else {
+            $node = $this->staticPaths[$regexPathSegment] ??= self::createStaticRoute($currentPathSegment);
+        }
+
+        $node->addPath($pathSegments, $markedRoute);
+    }
+
+    public static function convertDynamicSegmentToRegex(string $uriPart): string
+    {
+        $regex = '#\{'. Route::ROUTE_PARAM_NAME_REGEX . Route::ROUTE_PARAM_CUSTOM_REGEX .'\}#';
+
+        return preg_replace_callback(
+            $regex,
+            static fn ($matches) => trim($matches[2] ?? Route::DEFAULT_MATCHING_GROUP),
+            $uriPart,
+        );
+    }
+
+    private function regexSegment(): string
+    {
+        return match($this->type) {
+            RouteTreeNodeType::Root => '^',
+            RouteTreeNodeType::Static => "/{$this->segment}",
+            RouteTreeNodeType::Parameter => '/(' . $this->segment . ')',
+        };
+    }
+
+    public function toRegex(): string
+    {
+        $regexp = $this->regexSegment();
+
+        if ($this->staticPaths !== [] || $this->dynamicPaths !== []) {
+            $regexp .= "(?";
+
+            foreach ($this->staticPaths as $path) {
+                $regexp .= '|' . $path->toRegex();
```

Comment:
Won't the leading `|` be a problem?

Reply:
Ahh good question.
`(?|a|b|c)` is a valid regex. the `(?|` group is a "Branch Reset Group" Basically a switch statement in the regex. It also allows capture groups in the `|` with the same index as all others , which is really nice for us since all parameters will match on the same index everytime.

Read more about them here https://www.regular-expressions.info/branchreset.html

I didn't know about this this until I saw it in the sample implementation in the symfony router linked in #554

Reply:
Ok, maybe add this explanation in the code itself as well for clarity 👍

Reply:
done